File: 프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    int n=progresses.size();
    vector<int> s(n,0);
    for(int i=0;i<n;i++){
        int l=(100-progresses[i]+speeds[i]-1)/speeds[i];
        s[i]=l;
    }
    int count=1;
    int standard=s[0];
    for(int j=1;j<n;j++){
        if(standard>=s[j]){
            count++;
        }
        else{
            answer.push_back(count);
            standard=s[j];
            count=1;
        }
    }
    answer.push_back(count);
    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {

    public int[] solution(int[] progresses, int[] speeds) {

        ArrayList<Integer> list = new ArrayList<>();

        Stack<Integer> stck = new Stack<>();

        // 작업 완료일까지 계산
        for (int i = progresses.length - 1; i >= 0; i--) {

            int spend = (int)Math.ceil(
                (double)(100 - progresses[i]) / speeds[i]
            );

            stck.push(spend);
        }

        while (!stck.isEmpty()) {

            int standard = stck.pop();

            int count = 1;

            while (!stck.isEmpty() && stck.peek() <= standard) {
                stck.pop();
                count++;
            }

            list.add(count);
        }

        // ArrayList -> int[]
        int[] answer = new int[list.size()];

        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }

        return answer;
    }
}